src/android/CameraPreviewFragment.java

```diff
@@ -114,47 +136,51 @@ public void startCamera() {
 
-        Size targetResolution = null;
-        if (targetSize > 0) {
-            targetResolution = CameraPreviewFragment.calculateResolution(getContext(), targetSize);
+        if (startCameraCallback != null) {
+            startCameraCallback.onCameraStarted(null);
         }
+    }
 
-        preview = new Preview.Builder().build();
-        imageCapture = new ImageCapture.Builder()
-                .setTargetResolution(targetResolution)
-                .build();
+    @SuppressLint("RestrictedApi")
+    public void deviceHasUltraWideCamera(HasUltraWideCameraCallback hasUltraWideCameraCallback) {
+        ListenableFuture<ProcessCameraProvider> cameraProviderFuture = ProcessCameraProvider.getInstance(getActivity());
+        ProcessCameraProvider cameraProvider = null;
 
-        cameraProvider.unbindAll();
         try {
-            camera = cameraProvider.bindToLifecycle(
-                    this,
-                    cameraSelector,
-                    preview,
-                    imageCapture
-            );
-        } catch (IllegalArgumentException e) {
-            // Error with result in capturing image with default resolution
+            cameraProvider = cameraProviderFuture.get();
+        } catch (ExecutionException | InterruptedException e) {
+            Log.e(TAG, "Error occurred while trying to obtain the camera provider: " + e.getMessage());
             e.printStackTrace();
-            imageCapture = new ImageCapture.Builder()
-                    .build();
-            camera = cameraProvider.bindToLifecycle(
-                    this,
-                    cameraSelector,
-                    preview,
-                    imageCapture
-            );
+            hasUltraWideCameraCallback.onResult(false);
+            return;
+        }
+        List<CameraInfo> cameraInfos = cameraProvider.getAvailableCameraInfos();
+
+        boolean defaultCamera = false;
+        boolean ultraWideCamera = false;
+        List<Camera2CameraInfoImpl> backCameras = new ArrayList<>();
+        for (CameraInfo cameraInfo : cameraInfos) {
+            if (cameraInfo instanceof Camera2CameraInfoImpl) {
+                Camera2CameraInfoImpl camera2CameraInfo = (Camera2CameraInfoImpl) cameraInfo;
+                if (camera2CameraInfo.getLensFacing() == CameraSelector.LENS_FACING_BACK) {
+                    backCameras.add(camera2CameraInfo);
+                }
+            }
         }
 
-        preview.setSurfaceProvider(viewFinder.getSurfaceProvider());
-
-        if (startCameraCallback != null) {
-            startCameraCallback.onCameraStarted(null);
+        for (Camera2CameraInfoImpl backCamera : backCameras) {
+            if (backCamera.getCameraCharacteristicsCompat().get(CameraCharacteristics.LENS_INFO_AVAILABLE_FOCAL_LENGTHS)[0] >= 2.4) {
+                defaultCamera = true;
+            } else if( backCamera.getCameraCharacteristicsCompat().get(CameraCharacteristics.LENS_INFO_AVAILABLE_FOCAL_LENGTHS)[0] < 2.4) {
```

Comment:
```suggestion
            } else if( backCamera.getCameraCharacteristicsCompat().get(CameraCharacteristics.LENS_INFO_AVAILABLE_FOCAL_LENGTHS)[0] < 2.4) {
```

For the 2.4 focal length, is there any doc?

Reply:
@zfir i can not find any official documentation

but this might help: https://www.xlrsecurity.com/blog/cctv-camera-lens-comparison-2-8mm-vs-4-0mm/#:~:text=In%20this%20case%2C%20the%20amount,results%20in%20a%20narrower%20shot.

maybe we can increase it to 2.8?? @dinitri

Reply:
@dinitri @zfir can i publish this?